Add text alignment keybindings

justbur opened the pull request:
I added some keybindings for the `xa` prefix in the spacemacs layer related to aligning text. I tried to guess the most common delimiters and create functions for those directly. Let me know if you want me to add or subtract from these. All of the functions repeat until the end of the line (unlike the default behavior of `align-regexp`).

Also, I couldn't figure out any logical organization to the spacemacs/funcs.el file so I added the new functions to the end.

TheBB commented:
Yes, please!

TheBB commented:
I made [some changes](https://github.com/TheBB/spacemacs/commit/99c26bc64432bdf9b73c6c42d0e51cb8cf70ad5f) to your code, because I'm really excited about this functionality. Let me know what you think! I added an optional `after` argument to `align-repeat` that switches the order of the regexp and the whitespace around. I also added a prefix argument to the generated functions that switches the before/after functionality. I made comma and semicolon default to 'after' because I think that is the tendency for those cases. I also added a colon command, which I think should be there, also defaulting to 'after'. (And I created bindings for colon and the parens).

I'm not sure what the decimal alignment is supposed to do. I didn't have much luck with it. Can you show a use case?

Here's the default behaviour after my changes for commas, semicolon, colon, equal sign, operators, ampersand, pipe and left paren. The left side is before and the right side is after. I think this coincides with the most common use cases and expected behaviours?

![](http://i.imgur.com/tqQGKC9.png)

justbur commented:
@TheBB Thank you! I incorporated your changes and put everything in one commit for simplicity (I'll break it out again if you don't like that). I think you're right about the default for ":". I'm not sure what the default should be for "," but your way is fine with me. 

For `align-repeat-decimal` I modified the answer from http://emacs.stackexchange.com/questions/47/align-vertical-columns-of-numbers-on-the-decimal-point. Sometimes it's nice for a table full of numbers to align according to the decimal points. There are ways to do this in latex tables for scientific papers for example. Here's some sample text

```
a $ 1.2 $ 4.00 5.1 6. 80.0
b $ 12.3 $ 45.6 7.89 10. 0.987
```

which becomes

```
a $  1.2 $  4.00  5.1    6.  80.0
b $ 12.3 $ 45.6   7.89  10.   0.987
```

I just thought it was nice to have since it's complicated to get working right. Maybe not the most useful command ever.

TheBB commented:
That's fine :+1:

Your example for digits work fine for me, but I see strange behaviour otherwise.
 
    102.9 $ 4.1 $ 843.123
    432.12 $ 43287.123 $ 454932.1

becomes

    102.9                 $                  4.1  $   843.123 
                                           432.12 $ 43287.123 $ 454932.1

I don't understand why, either.

tuhdo commented:
@TheBB If you want something automatic in a single command, you should take a look at [align-cols](http://www.emacswiki.org/emacs/align.el).

tuhdo commented:
I've just tried `align-cols` with all the example above and it works for all.

justbur commented:
@tuhdo `align-cols` doesn't align to decimal places. It's just using whitespace. For example, in

```
alpha, bravo charlie,
delta echo, foxtrot
```
`align-cols` won't align using the comma. `align-repeat` with "," gives 

```
alpha      , bravo charlie ,
delta echo , foxtrot
```

which at least lines up into meaningful columns. @TheBB in this example the current "xa," gives 

```
alpha,      bravo charlie,
                           delta echo, foxtrot
```

so it's somehow messing up the first column.

justbur commented:
That being said, `align-cols` might be a better default for "xaa".

justbur commented:
@TheBB Fixed the `align-repeat-decimal` problem. There is some problem with the regexp matching the newline characters, so I switched to only matching on tabs and spaces.

TheBB commented:
Thanks, it works now. Great work!

syl20bnr commented:
WOW you guys are amazing :+1: 
Thank you all !
Cherry-picked into `develop` branch, you can safely delete your branch.
